common/lib/capa/capa/responsetypes.py

```diff
@@ -720,14 +726,103 @@ def assign_choice_names(self):
 
     def get_score(self, student_answers):
 
+        """
+        Answer lists:
+          all_choices: the full list of checkboxes
+          student_answer: what the student actually chose (note no "s")
+          student_non_answer: what they didn't choose
+          self.correct_choices: boxes that should be checked
+          self.incorrect_choices: boxes that should NOT be checked
+        """
+        
+        # pipe for union of sets
+        all_choices = self.correct_choices | self.incorrect_choices   
+        
         student_answer = student_answers.get(self.answer_id, [])
 
         if not isinstance(student_answer, list):
             student_answer = [student_answer]
 
         no_empty_answer = student_answer != []
+        # "None apply" should really be a valid choice for "check all that apply",
+        # but it throws an error if all the checks are blank.
+        
         student_answer = set(student_answer)
 
+        student_non_answers = all_choices - student_answer
+        
+        if not no_empty_answer:
+            print "Empty Answer"
+            return CorrectMap(self.answer_id, 'incorrect')
+        
+        
+        """
+        This below checks to see whether we're using an alternate grading scheme.
+          Set partial_credit="false" (or remove it) to require an exact answer for any credit.
+          Set partial_credit="EDC" to count each choice for equal points (Every Decision Counts).
+          Set partial_credit="halves" to take half credit off for each error.
+        """
+        
+        tree = self.xml
+        problem_xml = tree.xpath('.')
+
+        # Partial credit type - only one type at a time right now.
+        credit_type = problem_xml[0].get('partial_credit', default=False)
+
+        try:
+            credit_type = str(credit_type).lower().strip()
+        except ValueError:
+            _ = self.capa_system.i18n.ugettext
+            # Translators: 'partial_credit' is an attribute name and should not be translated.
+            msg = _("partial_credit value should be one of 'EDC', 'halves', or 'false'.")
+            raise LoncapaProblemError(msg)
+            
+        if credit_type == 'false':
+        
+            pass
+
+        elif credit_type == 'halves':
+        
+            halves_error_count = 0
+            
+            for answer in student_answer:
+                if answer in self.incorrect_choices:
+                    halves_error_count += 1
+            for answer in student_non_answers:
+                if answer in self.correct_choices:
+                    halves_error_count += 1
```

Comment:
You don't have to use my code. I had to read your code a few times to understand what the loops were doing. I think there are more consise ways of doing things rather than two loops, but if you want to keep your code just add some brief comments explaining what you're doing, eg
`# Sum up the incorrect answers (wrong choices that were selected)`
`# Sum up the answers that were missed (right choices that weren't selected)`